plugin-server/src/main/ingestion-queues/kafka-queue.ts

```diff
@@ -51,7 +59,64 @@ export class KafkaQueue implements Queue {
         }
     }
 
-    private async eachBatch({
+    private async eachMessageBuffer(
+        message: KafkaMessage,
+        resolveOffset: EachBatchPayload['resolveOffset']
+    ): Promise<void> {
+        const bufferEvent = JSON.parse(message.value!.toString())
+        await runInstrumentedFunction({
+            server: this.pluginsServer,
+            event: bufferEvent,
+            func: (_) => this.workerMethods.ingestBufferEvent(bufferEvent),
+            statsKey: `kafka_queue.ingest_buffer_event`,
+            timeoutMessage: 'After 30 seconds still running ingestBufferEvent',
+        })
+        resolveOffset(message.offset)
+    }
+
+    private async eachBatchBuffer({ batch, resolveOffset, commitOffsetsIfNecessary }: EachBatchPayload): Promise<void> {
+        if (batch.messages.length === 0) {
+            return
+        }
+        const batchStartTimer = new Date()
+
+        const promises = []
+        let consumerSleep = 0
+        for (const message of batch.messages) {
+            // kafka timestamps are unix timestamps in string format
+            const processAt = Number(message.timestamp) + this.pluginsServer.BUFFER_CONVERSION_SECONDS * 1000
+            const delayUntilTimeToProcess = processAt - new Date().getTime()
+
+            if (delayUntilTimeToProcess < 0) {
+                promises.push(this.eachMessageBuffer(message, resolveOffset))
+            } else {
+                consumerSleep = Math.max(consumerSleep, delayUntilTimeToProcess)
+            }
+        }
+
+        await Promise.all(promises)
```

Comment:
we should commit the offsets right after this, currently we do the work, then try to set up the pausing, then commit offsets, if we die while sleeping we'd try to process them again & we should try to minimize the amount of time after processing something and committing offsets.

Reply:
Tagged you in another comment.

Also, we don't synchronously sleep. `setTimeout` runs in the background

Reply:
sure based on all you say it should work, though I'd find it easier to read if we had `await commitOffsetsIfNecessary()` right after this line unless there's a reason not to?

Reply:
There is a good reason - if I run `commitOffsetsIfNecessary` it will move the offset past this entire batch. We rely on `throw new DelayProcessing()` to commit only the messages we processed via `resolveOffset` and not the entire batch. 

I should however do some thinking about the problems of running `Promise.all` with this. This might be more robust if done in order.

Reply:
Yeah I think the way to go is to do these in order and set `partitionsConsumedConcurrently`